Add Vulkan op Add.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46900 Add Vulkan op Conv2D.
* **#44017 Add Vulkan op Add.**
* #46481 Add Vulkan tensor copy.

Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

Commit:
Add Vulkan op Add.

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."




[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/ops/Add.cpp:
```diff
@@ -0,0 +1,133 @@
+namespace {
+
+Tensor add_scalar(
+    const Tensor& self_,
```
Nit: usually I saw in pytorch code base that member fields are named with underscore at the end

AshkanAliabadi replied:
No that's a good idea.  I'll follow up with the same convention.  I'll use self_arg, etc.  I am not really sue whether we use camelCase or snake_case in PyTorch.  It seems to be a mix.  I can find instances of both even under native.

IvanKobzarev commented on aten/src/ATen/native/vulkan/ops/Add.cpp:
```diff
@@ -0,0 +1,133 @@
+          context->resource().pool.uniform(block).object);
+    }
+    else {
+      // TODO: Convert all to buffer and dispatch a buffer-only kernel.
```
Hm. Do you try to avoid on demand on conversion to image view to have them only explicitly done?

AshkanAliabadi commented:
tensor.has_image means can have image or tensor can be represented as an
image. If false, tensor cannot be represented as an image.

On Wed, Oct 21, 2020, 12:16 PM Ivan Kobzarev <notifications@github.com>
wrote:

> *@IvanKobzarev* commented on this pull request.
> ------------------------------
>
> In aten/src/ATen/native/vulkan/ops/Add.cpp
> <https://github.com/pytorch/pytorch/pull/44017#discussion_r509601805>:
>
> > +          v_output.extents(),
> +          // Write-only access bypasses synchronization but inserts appropriate
> +          // barriers if necessary.
> +          v_output.image(command_buffer, vTensor::Access::Write),
> +          // Read-only access is implied on const tensors and triggers an async
> +          // synchronization if necessary.
> +          v_self.image(command_buffer),
> +          // Read-only access is implied on const tensors and triggers an async
> +          // synchronization if necessary.
> +          v_other.image(command_buffer),
> +          // Object lifetime is managed by the resource pool.
> +          // It is OK not to keep track of the handle.
> +          context->resource().pool.uniform(block).object);
> +    }
> +    else {
> +      // TODO: Convert all to buffer and dispatch a buffer-only kernel.
>
> Hm. Do you try to avoid on demand on conversion to image view to have them
> only explicitly done?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/44017#pullrequestreview-514101333>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKAGYKW5NEAZSCXPBHDSL4XPZANCNFSM4QSR5H4A>
> .
>

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]

Commit:
Update on "Add Vulkan op Add."


Differential Revision: [D23820826](https://our.internmc.facebook.com/intern/diff/D23820826)

[ghstack-poisoned]
